quota/cache/cache.go

```diff
@@ -0,0 +1,200 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package cache contains a caching quota.Manager implementation.
+package cache
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian/quota"
+)
+
+// now is used in place of time.Now to allow tests to take control of time.
+var now = time.Now
+
+type manager struct {
+	qm                    quota.Manager
+	batchSize, maxEntries int
+
+	// mu guards cache
+	mu    sync.Mutex
+	cache map[quota.Spec]*bucket
+
+	// evictWg tracks evict() goroutines.
+	evictWg sync.WaitGroup
+}
+
+type bucket struct {
+	tokens       int
+	lastModified time.Time
+}
+
+// NewCachedManager wraps a quota.Manager with an implementation that caches tokens locally.
+//
+// batchSize determines the minimum number of tokens requested from qm for each GetTokens() request.
+// More tokens than batchSize may be requested in order to fulfill the present request.
+//
+// maxEntries determines the maximum number of cache entries, apart from global quotas. The oldest
+// entries are evicted as necessary, their tokens replenished via PutTokens() to avoid excessive
+// leakage.
+func NewCachedManager(qm quota.Manager, batchSize, maxEntries int) quota.Manager {
+	return &manager{
+		qm:         qm,
+		batchSize:  batchSize,
+		maxEntries: maxEntries,
+		cache:      make(map[quota.Spec]*bucket),
+	}
+}
+
+func (m *manager) GetUser(ctx context.Context, req interface{}) string {
+	return m.qm.GetUser(ctx, req)
+}
+
+func (m *manager) PeekTokens(ctx context.Context, specs []quota.Spec) (map[quota.Spec]int, error) {
+	return m.qm.PeekTokens(ctx, specs)
+}
+
+func (m *manager) PutTokens(ctx context.Context, numTokens int, specs []quota.Spec) error {
+	return m.qm.PutTokens(ctx, numTokens, specs)
+}
+
+func (m *manager) ResetQuota(ctx context.Context, specs []quota.Spec) error {
+	return m.qm.ResetQuota(ctx, specs)
+}
+
+func (m *manager) GetTokens(ctx context.Context, numTokens int, specs []quota.Spec) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Verify which buckets need more tokens, if any
+	specsToRefill := []quota.Spec{}
+	for _, spec := range specs {
+		bucket, ok := m.cache[spec]
+		if !ok || bucket.tokens < numTokens {
+			specsToRefill = append(specsToRefill, spec)
+		}
+	}
+
+	// Request the required number of tokens and add them to buckets
+	if len(specsToRefill) != 0 {
+		defer func() {
+			// Do not hold GetTokens on eviction, it won't change the result.
+			m.evictWg.Add(1)
+			go func() {
+				m.evict(ctx)
+				m.evictWg.Done()
+			}()
+		}()
+
+		// A more accurate count would be numTokens+m.batchSize-bucket.tokens, but that might force
+		// us to make a GetTokens call for each spec. A single call is likely to be more efficient.
+		tokens := numTokens + m.batchSize
+		if err := m.qm.GetTokens(ctx, tokens, specsToRefill); err != nil {
```

Comment:
Happy to keep `m.mu` locked for the time it takes to get a response to an RPC? That seems like a long time to keep a mutex locked, and blocks additional `GetTokens()` calls in the meantime, but I can see how you could end up making lots of duplicate GetTokens requests otherwise.